modules/pubmaticBidAdapter.js

```diff
@@ -2,137 +2,216 @@ var utils = require('src/utils.js');
+    'dctr': 'dctr', // Custom Targeting
+    'wiid': 'wiid', // Wrapper Impression ID
+    'profId': 'profId', // Legacy: Profile ID
+    'verId': 'verId', // Legacy: version ID
+    'pmzoneid': { // Zone ID
+      n: 'pmZoneId',
+      m: function(zoneId) {
+        if (utils.isStr(zoneId)) {
+          return zoneId.split(',').slice(0, 50).join();
+        } else {
+          return undefined;
+        }
+      }
+    }
+  };
+
+  function _initConf() {
+    var conf = {};
+    var currTime = new Date();
+
+    conf.SAVersion = '1100';
+    conf.wp = 'PreBid';
+    conf.js = 1;
+    conf.wv = constants.REPO_AND_VERSION;
+    _secure && (conf.sec = 1);
+    conf.screenResolution = screen.width + 'x' + screen.height;
+    conf.ranreq = Math.random();
+    conf.inIframe = window != top ? '1' : '0';
+
+    // istanbul ignore else
+    if (window.navigator.cookieEnabled === false) {
+      conf.fpcd = '1';
     }
 
-    // Load pubmatic script in an iframe, because they call document.write
-    _getBids();
+    try {
+      conf.pageURL = window.top.location.href;
+      conf.refurl = window.top.document.referrer;
+    } catch (e) {
+      conf.pageURL = window.location.href;
+      conf.refurl = window.document.referrer;
+    }
+
+    conf.kltstamp = currTime.getFullYear() +
+      '-' + (currTime.getMonth() + 1) +
+      '-' + currTime.getDate() +
+      ' ' + currTime.getHours() +
+      ':' + currTime.getMinutes() +
+      ':' + currTime.getSeconds();
+    conf.timezone = currTime.getTimezoneOffset() / 60 * -1;
+
+    return conf;
   }
 
-  function _getBids() {
-    // create the iframe
-    iframe = utils.createInvisibleIframe();
+  function _handleCustomParams(params, conf) {
+    // istanbul ignore else
+    if (!conf.kadpageurl) {
+      conf.kadpageurl = conf.pageURL;
+    }
 
-    var elToAppend = document.getElementsByTagName('head')[0];
+    var key, value, entry;
+    for (key in customPars) {
+      // istanbul ignore else
+      if (customPars.hasOwnProperty(key)) {
+        value = params[key];
+        // istanbul ignore else
+        if (value) {
+          entry = customPars[key];
+
+          if (typeof entry === 'object') {
+            value = entry.m(value, conf);
+            key = entry.n;
+          } else {
+            key = customPars[key];
+          }
+
+          if (utils.isStr(value)) {
+            conf[key] = value;
+          } else {
+            utils.logWarn('PubMatic: Ignoring param key: ' + customPars[key] + ', expects string-value, found ' + typeof value);
+          }
+        }
+      }
+    }
+    return conf;
+  }
 
-    // insert the iframe into document
-    elToAppend.insertBefore(iframe, elToAppend.firstChild);
+  function _cleanSlot(slotName) {
+    // istanbul ignore else
+    if (utils.isStr(slotName)) {
+      return slotName.replace(/^\s+/g, '').replace(/\s+$/g, '');
+    }
+    return '';
+  }
 
+  function _legacyExecution(conf, slots) {
+    var url = _generateLegacyCall(conf, slots);
```

Comment:
I'm not clear why this is still loaded into an iframe. Can you not use the `ajax` method here?

Reply:
Our server response uses global namespace also our server is not supporting CORS.
By calling our server from a friendly-iframe we can get better priority than calling from a script tag thus we are calling our server using an iframe.